File: src/ui_gtk/Envelope.py
# coding=utf-8


# Copyright 2008 Tomi Jylhä-Ollila
#
# This file is part of Kunquat.
#
# Kunquat is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Kunquat is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Kunquat.  If not, see <http://www.gnu.org/licenses/>.


import pygtk
pygtk.require('2.0')
import gtk, gobject, cairo, pango
from gtk import gdk
import math

import liblo


class Envelope(gtk.Widget):

	def handle_key(self, widget, event):
		pass

	def handle_focus_in(self, widget, event):
		pass

	def handle_focus_out(self, widget, event):
		pass

	def handle_button(self, widget, event):
		pass

	def handle_motion(self, widget, event):
		x = event.x - self.ruler_width
		y = self.margin_top + self.view_height - event.y
		self.pointer_coords = (x / self.zoom_x, y / self.zoom_y)
		self.queue_draw()
		return True

	def do_realize(self):
		self.set_flags(self.flags()
				| gtk.REALIZED
				| gtk.SENSITIVE
				| gtk.PARENT_SENSITIVE
				| gtk.CAN_FOCUS)
		self.window = gdk.Window(
				self.get_parent_window(),
				width = self.allocation.width,
				height = self.allocation.height,
				window_type = gdk.WINDOW_CHILD,
				wclass = gdk.INPUT_OUTPUT,
				event_mask = self.get_events()
						| gdk.EXPOSURE_MASK
						| gdk.KEY_PRESS
						| gdk.BUTTON_PRESS_MASK
						| gdk.POINTER_MOTION_MASK
						| gdk.POINTER_MOTION_HINT_MASK
						| gdk.FOCUS_CHANGE)
		self.window.set_user_data(self)
		self.style.attach(self.window)
		self.style.set_background(self.window, gtk.STATE_NORMAL)
		self.window.move_resize(*self.allocation)
		self.connect('key-press-event', self.handle_key)
		self.connect('focus-in-event', self.handle_focus_in)
		self.connect('focus-out-event', self.handle_focus_out)
		self.connect('button-press-event', self.handle_button)
		self.connect('motion-notify-event', self.handle_motion)

	def do_unrealize(self):
		self.window.destroy()
		self.window.set_user_data(None)

	def do_size_request(self, requisition):
		if not self.scale_y and self.view_height <= 128:
			requisition.height = self.view_height + self.ruler_height + self.margin_top
		if not self.scale_x and self.view_width <= 256:
			requisition.width = self.view_width + self.ruler_width + self.margin_right

	def do_size_allocate(self, allocation):
		self.allocation = allocation
		if self.flags() & gtk.REALIZED:
			self.window.move_resize(*allocation)

	def do_expose_event(self, event):
		cr = self.window.cairo_create()
		cr.rectangle(event.area.x, event.area.y,
				event.area.width, event.area.height)
		cr.clip()
		self.draw(cr, *self.window.get_size())

	def draw(self, cr, width, height):
		cr.set_source_rgb(*self.ptheme['Background colour'])
		cr.rectangle(0, 0, width, height)
		cr.fill()
		cr.set_line_width(1)

		max_x = 0
		max_y = 0
		for node in self.nodes:
			if max_x < abs(node[0]):
				max_x = abs(node[0])
			if max_y < abs(node[1]):
				max_y = abs(node[1])

		view_width = width - self.ruler_width - self.margin_right
		view_height = height - self.ruler_height - self.margin_top
		if self.scale_x and max_x > (view_width + 1) / self.zoom_x:
			self.zoom_x = 2**math.floor(math.log((view_width + 1) / max_x, 2))
		if self.scale_y and max_y > (view_height + 1) / self.zoom_y:
			self.zoom_y = 2**math.floor(math.log((view_height + 1) / max_y, 2))

		coords_str = '---'
		if self.pointer_coords:
			coords_str = '(%f, %f)' % self.pointer_coords
		pl = self.create_pango_layout(coords_str)
		pl.set_font_description(self.ruler_font)
		w, h = pl.get_size()
		w //= pango.SCALE
		h //= pango.SCALE
		cr.set_source_rgb(*self.ptheme['Node colour'])
		cr.move_to(width - w - self.margin_right, self.margin_top)
		cr.update_layout(pl)
		cr.show_layout(pl)

		self.draw_ruler_y(cr, width, height)
		self.draw_ruler_x(cr, width, height)

		cr.set_source_rgb(*self.ptheme['Line colour'])
		prev = None
		for node in self.nodes:
			if prev:
				prev_y = self.margin_top + self.view_height - (prev[1] * self.zoom_y)
				prev_x = self.ruler_width + (prev[0] * self.zoom_x)
				y = self.view_height - (node[1] * self.zoom_y)
				x = (node[0] * self.zoom_x)
				cr.move_to(prev_x, prev_y)
				cr.rel_line_to(x, y)
				cr.stroke()
			prev = node

		for node in self.nodes:
			self.draw_node(cr, self.ruler_width + (node[0] * self.zoom_x) + 0.5,
					self.margin_top + self.view_height -
					(node[1] * self.zoom_y) - 0.5,
					False)

	def draw_ruler_y(self, cr, width, height):
		cr.set_source_rgb(*self.ptheme['Border colour'])
		cr.move_to(self.ruler_width + 0.5, self.margin_top)
		cr.rel_line_to(0, self.view_height)
		cr.stroke()

	def draw_ruler_x(self, cr, width, height):
		cr.set_source_rgb(*self.ptheme['Border colour'])
		cr.move_to(self.ruler_width, self.view_height + self.margin_top - 0.5)
		cr.rel_line_to(self.view_width, 0)
		cr.stroke()

	def draw_node(self, cr, x, y, selected):
		if selected:
			cr.set_source_rgb(*self.ptheme['Node highlight colour'])
		else:
			cr.set_source_rgb(*self.ptheme['Node colour'])
		cr.rectangle(x - self.node_width / 2, y - self.node_width / 2,
				self.node_width, self.node_height)
		cr.fill()

	def do_redraw(self):
		if self.window:
			alloc = self.get_allocation()
			rect = gdk.Rectangle(alloc.x, alloc.y,
					alloc.width, alloc.height)
			self.window.invalidate_rect(rect, True)
			self.window.process_updates(True)
		return True

	def __init__(self, engine, server, song_id,
			x_desc, y_desc):
		gtk.Widget.__init__(self)
		self.engine = engine
		self.server = server

		self.ptheme = {
			'Ruler font': 'Sans 8',
			'Background colour': (0, 0, 0),
			'Border colour': (0.6, 0.6, 0.6),
			'Ruler bg colour': (0, 0.1, 0.2),
			'Ruler fg colour': (0.7, 0.7, 0.7),
			'Line colour': (0.7, 0.5, 0.4),
			'Node colour': (1, 0.9, 0.8),
			'Node highlight colour': (1, 1, 1),
		}

		self.ruler_font = pango.FontDescription(self.ptheme['Ruler font'])
		self.ruler_font_size = self.ruler_font.get_size() // pango.SCALE
		self.ruler_width = self.ruler_font_size * 5
		self.ruler_height = self.ruler_font_size * 3

		self.min_x, self.max_x, self.step_x, self.scale_x = x_desc
		self.min_y, self.max_y, self.step_y, self.scale_y = y_desc

		self.zoom_x = 1.0 / self.step_x
		self.zoom_y = 1.0 / self.step_y

		self.view_height = (self.max_y - self.min_y) // self.step_y
		self.view_width = (self.max_x - self.min_x) // self.step_x

		self.margin_top = 8
		self.margin_right = 8

		self.node_width = 3.0
		self.node_height = 3.0

		self.nodes = []

		self.pointer_coords = None


gobject.type_register(Envelope)

File: src/core/Envelope.c


/*
 * Copyright 2009 Tomi Jylhä-Ollila
 *
 * This file is part of Kunquat.
 *
 * Kunquat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Kunquat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Kunquat.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <stdlib.h>
#include <assert.h>
#include <stdio.h>
#include <math.h>

#include <Envelope.h>

#include <xmemory.h>


struct Envelope
{
    double min_x;
    double max_x;
    double step_x;
    double min_y;
    double max_y;
    double step_y;
    bool first_x_locked;
    bool first_y_locked;
    bool last_x_locked;
    bool last_y_locked;
    Envelope_int interp;
    int node_count;
    int nodes_max;
    int nodes_res;
    int marks[ENVELOPE_MARKS_MAX];
    double* nodes;
};


Envelope* new_Envelope(int nodes_max,
        double min_x, double max_x, double step_x,
        double min_y, double max_y, double step_y)
{
    assert(nodes_max > 1);
    assert(!isnan(min_x));
    assert(!isnan(max_x));
    assert(step_x >= 0);
    assert(!isnan(min_y));
    assert(!isnan(max_y));
    assert(step_y >= 0);
    Envelope* env = xalloc(Envelope);
    if (env == NULL)
    {
        return NULL;
    }
    env->nodes = xnalloc(double, nodes_max * 2);
    if (env->nodes == NULL)
    {
        xfree(env);
        return NULL;
    }
    env->node_count = 0;
    env->nodes_max = nodes_max;
    env->nodes_res = nodes_max;
    for (int i = 0; i < ENVELOPE_MARKS_MAX; ++i)
    {
        env->marks[i] = -1;
    }
    env->interp = ENVELOPE_INT_LINEAR;
    env->min_x = min_x;
    env->max_x = max_x;
    env->step_x = step_x;
    env->min_y = min_y;
    env->max_y = max_y;
    env->step_y = step_y;
    env->first_x_locked = false;
    env->first_y_locked = false;
    env->last_x_locked = false;
    env->last_y_locked = false;
    return env;
}


int Envelope_node_count(Envelope* env)
{
    assert(env != NULL);
    return env->node_count;
}


void Envelope_set_interp(Envelope* env, Envelope_int interp)
{
    assert(env != NULL);
//  assert(interp >= ENVELOPE_INT_NEAREST);
    assert(interp < ENVELOPE_INT_LAST);
    env->interp = interp;
    return;
}


void Envelope_set_mark(Envelope* env, int index, int value)
{
    assert(env != NULL);
    assert(index >= 0);
    assert(index < ENVELOPE_MARKS_MAX);
    env->marks[index] = value;
    return;
}


int Envelope_get_mark(Envelope* env, int index)
{
    assert(env != NULL);
    assert(index >= 0);
    assert(index < ENVELOPE_MARKS_MAX);
    return env->marks[index];
}


Envelope_int Envelope_get_interp(Envelope* env)
{
    assert(env != NULL);
    return env->interp;
}


int Envelope_set_node(Envelope* env, double x, double y)
{
    assert(env != NULL);
    assert(isfinite(x));
    assert(isfinite(y));
    if (env->node_count >= env->nodes_max)
    {
        assert(env->node_count == env->nodes_max);
        return -1;
    }
    if (env->node_count > 0)
    {
        if (x < env->nodes[0] && (env->first_x_locked || env->first_y_locked))
        {
            return -1;
        }
        if (x > env->nodes[env->node_count * 2 - 2]
                && (env->last_x_locked || env->last_y_locked))
        {
            return -1;
        }
    }
    int start = 0;
    int end = env->node_count - 1;
    while (start <= end)
    {
        int middle = (start + end) / 2;
        if (env->nodes[middle * 2] < x)
        {
            start = middle + 1;
        }
        else if (env->nodes[middle * 2] > x)
        {
            end = middle - 1;
        }
        else
        {
            Envelope_move_node(env, middle, x, y);
            return middle;
        }
    }
    ++env->node_count;
    for (int i = start + 1; i < env->node_count; ++i)
    {
        env->nodes[i * 2] = env->nodes[i * 2 - 2];
        env->nodes[i * 2 + 1] = env->nodes[i * 2 - 1];
    }
    env->nodes[start * 2] = x;
    env->nodes[start * 2 + 1] = y;
    for (int i = 0; i < ENVELOPE_MARKS_MAX; ++i)
    {
        if (env->marks[i] >= start)
        {
            ++env->marks[i];
        }
    }
    return start;
}


bool Envelope_del_node(Envelope* env, int index)
{
    assert(env != NULL);
    assert(index >= 0);
    if (index >= env->node_count)
    {
        return false;
    }
    if (index == 0 && (env->first_x_locked || env->first_y_locked))
    {
        return false;
    }
    if (index == env->node_count - 1
            && (env->last_x_locked || env->last_y_locked))
    {
        return false;
    }
    for (int i = index * 2; i < env->node_count * 2 - 2; i += 2)
    {
        env->nodes[i] = env->nodes[i + 2];
        env->nodes[i + 1] = env->nodes[i + 3];
    }
    --env->node_count;
    for (int i = 0; i < ENVELOPE_MARKS_MAX; ++i)
    {
        if (env->marks[i] > index)
        {
            --env->marks[i];
        }
    }
    return true;
}


double* Envelope_get_node(Envelope* env, int index)
{
    assert(env != NULL);
    assert(index >= 0);
    if (index >= env->node_count)
    {
        return NULL;
    }
    return env->nodes + index * 2;
}


double* Envelope_move_node(Envelope* env, int index, double x, double y)
{
    assert(env != NULL);
    assert(index >= 0);
    assert(isfinite(x));
    assert(isfinite(y));
    if (index >= env->node_count)
    {
        return NULL;
    }
    double max_x = env->max_x;
    double max_y = env->max_y;
    double min_x = env->min_x;
    double min_y = env->min_y;
    if (index == env->node_count - 1)
    {
        if (env->last_x_locked)
        {
            min_x = max_x = env->nodes[env->node_count * 2 - 2];
        }
        if (env->last_y_locked)
        {
            min_y = max_y = env->nodes[env->node_count * 2 - 1];
        }
    }
    else
    {
        double next_x = env->nodes[index * 2 + 2];
        max_x = next_x - env->step_x;
        if (max_x >= next_x)
        {
            max_x = nextafter(max_x, -INFINITY);
        }
    }
    if (index == 0)
    {
        if (env->first_x_locked)
        {
            min_x = max_x = env->nodes[0];
        }
        if (env->first_y_locked)
        {
            min_y = max_y = env->nodes[1];
        }
    }
    else
    {
        double prev_x = env->nodes[index * 2 - 2];
        min_x = prev_x + env->step_x;
        if (min_x <= prev_x)
        {
            min_x = nextafter(min_x, INFINITY);
        }
    }
    if (min_x > max_x)
    {
        return env->nodes + index * 2;
    }
    if (min_x > x)
    {
        env->nodes[index * 2] = min_x;
    }
    else if (max_x < x)
    {
        env->nodes[index * 2] = max_x;
    }
    else
    {
        env->nodes[index * 2] = x;
    }
    if (min_y > y)
    {
        env->nodes[index * 2 + 1] = min_y;
    }
    else if (max_y < y)
    {
        env->nodes[index * 2 + 1] = max_y;
    }
    else
    {
        env->nodes[index * 2 + 1] = y;
    }
    return env->nodes + index * 2;
}


double Envelope_get_value(Envelope* env, double x)
{
    assert(env != NULL);
    assert(isfinite(x));
    if (env->node_count == 0
            || x < env->nodes[0]
            || x > env->nodes[env->node_count * 2 - 2])
    {
        return NAN;
    }
    int start = 0;
    int end = env->node_count - 1;
    while (start <= end)
    {
        int middle = (start + end) / 2;
        if (env->nodes[middle * 2] < x)
        {
            start = middle + 1;
        }
        else if (env->nodes[middle * 2] > x)
        {
            end = middle - 1;
        }
        else
        {
            return env->nodes[middle * 2];
        }
    }
    assert(start < env->node_count);
    assert(end >= 0);
    assert(start == end + 1);
    double prev_x = env->nodes[end * 2];
    double prev_y = env->nodes[end * 2 + 1];
    double next_x = env->nodes[start * 2];
    double next_y = env->nodes[start * 2 + 1];
    switch (env->interp)
    {
        case ENVELOPE_INT_NEAREST:
        {
            if (x - prev_x < next_x - x)
            {
                return prev_y;
            }
            else
            {
                return next_y;
            }
        } break;
        case ENVELOPE_INT_LINEAR:
        {
            double interval = next_x - prev_x;
            double y = prev_y * (next_x - x) / interval
                    + next_y * (x - prev_x) / interval;
            return y;
        } break;
        default:
            assert(false);
    }
    assert(false);
    return NAN;
}


void Envelope_set_first_lock(Envelope* env, bool lock_x, bool lock_y)
{
    assert(env != NULL);
    assert(env->node_count > 0);
    env->first_x_locked = lock_x;
    env->first_y_locked = lock_y;
    return;
}


void Envelope_set_last_lock(Envelope* env, bool lock_x, bool lock_y)
{
    assert(env != NULL);
    assert(env->node_count > 0);
    env->last_x_locked = lock_x;
    env->last_y_locked = lock_y;
    return;
}


void del_Envelope(Envelope* env)
{
    assert(env != NULL);
    assert(env->nodes != NULL);
    xfree(env->nodes);
    xfree(env);
    return;
}


